### ISSUE TITLE
Improve hint enumeration performance

### ISSUE THREAD

**Issue by zsims** [zsims/hunt-and-peck#1]
Enumerating hints for "busy" windows can be quite slow, sometime taking ~3 seconds.

**Comment by zsims** [comment]
Can confirm that this is mostly in the way AutomationElement.FindAll is being used. It supports "caching" but only after the first visit :( https://msdn.microsoft.com/en-us/library/system.windows.automation.automationelement.findall%28v=vs.110%29.aspx

![hot](https://cloud.githubusercontent.com/assets/1764368/6382532/8ae01e76-bd9d-11e4-827e-02bcc2f7e489.PNG)

**Comment by oysteinkrog** [comment]
I've been playing with it a bit today but was unable to improve performance much, things I tried:
1. Switch to COM-based UI automation API (http://uiacomwrapper.codeplex.com). I think this gave a (very) slight performance improvement. It seems that the .net wrapper can be a bit slow for some cases.
2. Enumerate only certain control types (e.g. buttons), no real gain.
3. Enumerate ahead of time (when active window changes). I have not yet been able to get this to work as the UI automation API throws a TypeLoadException, will continue working on this as it seems the most promising.

**Comment by zsims** [comment]
Thanks a lot for contributing :smile:. I've also (albeit briefly) thought about your third option, does introduce some interesting challenges, the idea being to cache top level elements on [focused window change](https://msdn.microsoft.com/en-us/library/windows/desktop/dd373640%28v=vs.85%29.aspx)? Should work in theory :)

Feel free to assign this issue to yourself.

**Comment by zsims** [comment]
I tried another option, using the in-built cache functionality of the automation framework. With negative results :disappointed: 

For reference, 3x runs (in a row without re-launching), timing how long it took to create a hint session in `UiAutomationHintProviderService.cs`:

**Wordpad**
Before = 884 + 604 + 752 ms
After = 1092 + 996 + 945 ms

**Word**
Before = 519 + 604 + 547 ms
After = 755 + 677 + 701 ms

Will try your ahead-of-time idea when I can :+1:

**Comment by zsims** [comment]
More playing :dancers:. I whipped up your idea -- basically, build the `HintSession` ahead of time to mitigate the performance problems with the UIA API, see ec29786ebc7407d8ef6a83c4ecde7e6581d2eee2. It works quite well, assuming the hint session has already been built.

It's still WIP, but:

 1. Uses [Windows Events](https://msdn.microsoft.com/en-us/library/windows/desktop/dd318066%28v=vs.85%29.aspx) to detect when a new window becomes focused 
 2. Doesn't yet handle Windows with dynamic content that well
 3. Doesn't handle window resize/scroll
 4. Sometimes, it's not "ahead" enough. E.g. using Explorer, hints can take ~500ms to enumerate but I already know what I want to activate.

Cheers for the idea :smile:

**Comment by oysteinkrog** [comment]
Ah cool, I see that you invoke on the gui thread, that must be the fix for the exception I was struggling with! (I should have known:P)
I'm not sure if you looked at any of the wip code I have in my fork...
I did pretty much what you did: https://github.com/oysteinkrog/hunt-and-peck/commit/48d4da2712cb80466d8f533c4b9c0a90d9a96118
I decided to also replace the windows forms stuff with a low-level native window:
https://github.com/oysteinkrog/hunt-and-peck/commit/7965106fe749d781a81781b18a62aa4238ece6d5
I also played around with AccessKeys, I'm considering adding another mode/hotkey that shows the access keys instead of random keys, would be a good mode for learning to use them.

**Comment by zsims** [comment]
> I see that you invoke on the gui thread, that must be the fix for the exception I was struggling with! (I should have known:P)

Seemed to help, I'm not 100% sure if it's required. Bit rusty with COM.

> I'm not sure if you looked at any of the wip code I have in my fork...

Only just looked. Your stuff is much cleaner :+1: 

Agree re: AccessKeys, even just making sure the same label is applied to the same hint (or there abouts). It's a bit frustrating that the label changes each time for the same button.

I think the ahead-of-time hint session is a good idea, and seems to work, but not sure how to handle the case of windows that change -- short of digging into a lot more of the events. Any ideas here?

**Comment by zsims** [comment]
> Ah cool!, I see that you invoke on the gui thread, that must be the fix for the exception I was struggling with! (I should have known:P)

Hmm, in hind sight... I think this is wrong [according to MSDN](https://msdn.microsoft.com/en-us/library/ms788709%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396)

> Because of the way Microsoft UI Automation uses Windows messages, conflicts can occur when a client application attempts to interact with its own UI on the UI thread.
There is a brief Window when the app launches, and I suspect this was causing issues but the `Dispatcher.Invoke` essentially "delays" this (by putting a message on the UI event loop). Just a theory, but probably needs to be "on another thread" all together.

I've also been looking at the [UI automation events](https://msdn.microsoft.com/en-us/library/ms752297(v=vs.110).aspx), wonder if this is a good way to detect "changes" in Windows?

**Comment by zsims** [comment]
For reference, pretty good blog on improving UIA performance: https://blogs.msdn.microsoft.com/winuiautomation/2013/06/04/a-little-uia-qa/ (TL;DR: caching and the native COM API)

**Comment by zsims** [comment]
Switched to the native API, this has made a drastic improvement.

In Visual Studio (start page, average across 3 clean runs):
Before: ~2024ms 
After: ~616ms

Still room for improvement as this isn't using any of the UIA caching APIs.

**Comment by extratype** [comment]
I discovered enumeration takes much longer if ran on kind of.. background (tested on Windows 10). There's no official way to make it faster. I found a dirty hack (https://github.com/extratype/hunt-and-peck/commit/55c4377b4fda458c7c20c09480df9c323c53188b) but it's not always working.

**Comment by zsims** [comment]
Interesting... Do you know if there's any documentation in what impact this has? Didn't seem to make much of a difference for me. And https://msdn.microsoft.com/en-us/library/windows/desktop/ms685100\(v=vs.85\).aspx doesn't really mention anything about priority and CPU load (which I assume is the goal?)

**Comment by extratype** [comment]
I couldn't find any. I have searched MSDN and this forum (https://social.msdn.microsoft.com/Forums/windowsdesktop/en-US/home?forum=windowsaccessibilityandautomation). I just figured it out that UI Automation uses WM_GETOBJECT internally and there's no message priorities we can control. I tried adjusting process/thread priorities, but no gain.

I applied the commit https://github.com/extratype/hunt-and-peck/commit/55c4377b4fda458c7c20c09480df9c323c53188b to the current master branch, and the enumeration for Windows explorer took ~100ms less in worse cases. To reproduce, repeatedly press the hotkey with a short delay, and then with long delays. You will notice performance varies.

This matters in my branch (in hundreds of milliseconds), where I exploit UIA APIs to get only on-screen elements from an arbitrarily long list or tree view. It's much slower on average, but with the commit, it's barely tolerable.

**Comment by zsims** [comment]
Thanks for the investigation! It would be nice to get a consistent gain, even by enumerating hints ahead of time or changing the way the tree is walked. That being said, the switch to the native API has drastically improved things from what I've seen. Thoughts?

**Comment by extratype** [comment]
I agree on the ahead-of-time idea. You may try these [testing tools](https://msdn.microsoft.com/en-us/library/windows/desktop/dd373661.aspx) included in Windows Kits to inspect elements and monitor events and find something useful.
